cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    process::exit,
+};
+
+use anyhow::{Context, Result};
+use clap::Parser;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+    sync::oneshot,
+};
+use tracing::{debug, error, info, Instrument};
+
+#[derive(Parser)]
+#[clap(version)]
+struct Args {
+    #[clap(
+        long,
+        env = "LINKERD_VALIDATION_LOG_LEVEL",
+        default_value = "linkerd_cni_validator=info,warn"
+    )]
+    log_level: kubert::LogFilter,
+
+    #[clap(long, default_value = "plain")]
+    log_format: kubert::LogFormat,
+
+    #[clap(long, default_value = "0.0.0.0:4140")]
+    outbound_proxy_addr: SocketAddr,
+
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_HOST")]
+    target_ip: IpAddr,
+
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_PORT")]
+    target_port: u16,
+}
+
+static REDIRECT_RESPONSE: &str = "REDIRECTION SUCCESSFUL";
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    let Args {
+        log_level,
+        log_format,
+        outbound_proxy_addr,
+        target_ip,
+        target_port,
+    } = Args::parse();
+
+    log_format.try_init(log_level)?;
+    let target_addr = SocketAddr::new(target_ip, target_port);
+
+    info!(%outbound_proxy_addr, %target_addr, "Validating outbound traffic is redirected to proxy's outbound port");
+
+    let (shutdown_tx, shutdown_rx) = kubert::shutdown::sigint_or_sigterm()?;
+    let (ready_tx, ready_rx) = oneshot::channel::<Result<()>>();
+
+    let server_task = tokio::spawn(outbound_serve(outbound_proxy_addr, ready_tx, shutdown_rx));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(target_addr, ready_rx));
+
+    tokio::select! {
+        task = validation_task => {
+            let validation_result = task.expect("Failed to run validator task");
+            if let Err(err) = validation_result {
```

Comment:
nit: I'd probably omit the temp var

```suggestion
            if let Err(err) = task.expect("Failed to run validator task") {
```